File: javascript/102_-_Binary_Tree_Level_Order_Traversal.js
/**
 * 102. Binary Tree Level Order Traversal - Medium
 * Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
 var levelOrder = function(root) {
    let queue = [];
    let ans = [];
    
    if (root) {
        queue.push(root);
    }
    
    while (queue.length > 0) {
        let queueCount = queue.length;
        let level = [];
        
        for (let i = 0; i < queueCount; ++i) {
            let current = queue.shift();
            
            if (current.left) {
                queue.push(current.left);
            }
            
            if (current.right) {
                queue.push(current.right);
            }
            
            level.push(current.val);
        }
        
        ans.push(level);
    }
    
    return ans;
};

File: csharp/102_-_Binary_Tree_Level_Order_Traversal.cs
/**
 * 102. Binary Tree Level Order Traversal - Medium
 * Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        IList<IList<int>> ans = new List<IList<int>>();
        Queue<TreeNode> queue = new Queue<TreeNode>();
        
        if (root != null) {
            queue.Enqueue(root);
        }
        
        while (queue.Count > 0) {
            IList<int> currentLevel = new List<int>();
            int queueCount = queue.Count;
            
            for (int i = 0; i < queueCount; ++i) {
                TreeNode current = queue.Dequeue();
                
                if (current.left != null) {
                    queue.Enqueue(current.left);
                }
                
                if (current.right != null) {
                    queue.Enqueue(current.right);
                }
                
                currentLevel.Add(current.val);
            }
            
            ans.Add(currentLevel);
        }
        
        return ans;
    }
}